src/cargo/core/source_id.rs

```diff
@@ -468,34 +468,51 @@ impl SourceId {
 
     /// Creates a new `SourceId` from this source with the given `precise`.
     pub fn with_git_precise(self, fragment: Option<String>) -> SourceId {
-        SourceId::wrap(SourceIdInner {
-            precise: fragment.map(|f| Precise::GitUrlFragment(f)),
-            ..(*self.inner).clone()
-        })
+        let precise = fragment.map(|f| Precise::GitUrlFragment(f));
+        if self.inner.precise == precise {
+            self
+        } else {
+            SourceId::wrap(SourceIdInner {
+                precise,
+                ..(*self.inner).clone()
+            })
+        }
```

Comment:
Should we factor this out with a `with_precise` that takes an `Option` so it'd work in every case?

Reply:
One of `with_precise_from` or `with_git_precise` will need to do one extra clone. But it seams worth it.